### ISSUE TITLE
Miri doesn't understand C-variadic functions defined in Rust

### ISSUE THREAD

**Issue by bugadani** [rust-lang/miri#1892]
I have a program written in C and Rust. This program has a function that is variadic and written in C. To enable testing my Rust code, I needed to mock this function using the `#[feature(c_variadic)]` unstable feature.

This means I have not a function in my test code defined like the following:
```rust
pub unsafe extern "C" fn variadic_fn(
        some_arg: *const c_char,
        mut args: ...
    ) {
...
}
```

When running Miri on my test code, I'm greeted with the following error:
`error: Undefined Behavior: calling a function with argument of type std::ffi::VaListImpl passing data of type *const i8`

**Comment by RalfJung** [comment]
Yeah, Miri probably does not support varargs at all currently. I assume this is another thing that is blocked on https://github.com/rust-lang/rust/issues/56166.

If you could provide a self-contained example demonstrating the problem, that would be helpful as a testcase. :)

**Comment by bugadani** [comment]
Sure thing.

```rust
#![feature(c_variadic)]

use std::os::raw::c_char;

macro_rules! c_str {
    ($lit:expr) => {
        concat!($lit, "\0").as_ptr() as *const c_char
    };
}

#[allow(unused_unsafe)]
pub unsafe extern "C" fn variadic_fn(_: *const c_char, mut _args: ...) {
    todo!()
}

fn main() {
    unsafe {
        variadic_fn(c_str!("c string"), 1, 2, 3);
    }
}
```

**Comment by RalfJung** [comment]
FWIW, https://github.com/rust-lang/rust/pull/91342 should help with this.

But actually implementing `VaList` and the associated functions to retrieve the arguments would probably still require a bunch of work, fiddling with many horrible platform-specific details.

**Comment by RalfJung** [comment]
To be clear, I don't currently have plans to work on those platform-specific parts -- it will be a lot of work for very little gain, as such Rust-to-Rust calls of c-variadic functions are exceedingly rare, to my knowledge.

An alternative for your usecase might be to use `cfg(miri)` for mocking, so you can replace the c-variadic calls with regular Rust functions when testing under Miri.

**Comment by bugadani** [comment]
No worries, no worries, as a workaround I banished 99% of my C code, so this isn't important to me any more :)

**Comment by saethlin** [comment]
Just ran into this in the test suite of `kvm-ioctls`:
```
error: Undefined Behavior: incorrect number of arguments: got 2, expected 3
   --> src/ioctls/system.rs:71:28
    |
71  |         let ret = unsafe { open("/dev/kvm\0".as_ptr() as *const c_char, open_flags) };
    |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ incorrect number of arguments: got 2, expected 3
    |
    = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
    = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
    = note: inside `ioctls::system::Kvm::open_with_cloexec` at src/ioctls/system.rs:71:28
note: inside `ioctls::system::Kvm::new` at src/ioctls/system.rs:42:18
```

**Comment by RalfJung** [comment]
No, that's a different bug -- this bug here is about calling C variadic functions *that are defined in Rust*, whereas you are calling an FFI function that Miri emulates via a shim. It looks like our implementation of `open` only supports the 3-argument form. Could you open a new issue for this?

**Comment by RalfJung** [comment]
Ah, a PR has already been opened. :D
https://github.com/rust-lang/miri/pull/1970#issuecomment-1030733571

**Comment by saethlin** [comment]
Could Miri report this as an unsupported operation, not UB? Would that close this issue?

**Comment by RalfJung** [comment]
It does:

https://github.com/rust-lang/rust/blob/d7a2d9ae0e7e4b3c5811bdfd4809cfc772062140/compiler/rustc_const_eval/src/interpret/terminator.rs#L362

However, the issue you reported is separately detected before and considered UB:

https://github.com/rust-lang/rust/blob/d7a2d9ae0e7e4b3c5811bdfd4809cfc772062140/compiler/rustc_const_eval/src/interpret/terminator.rs#L357

Possibly we should remove that particular check. For our FFI implementation we ended up saying that we don't want to make a difference between variadic and non-variadic arguments, so this would make us more consistent. (But to my knowledge, at least on Apple's aarch64 targets, that is actually unsound.)

> Would that close this issue?

No, what would close this issue is actually getting [the testcase](https://github.com/rust-lang/miri/issues/1892#issuecomment-927654108) to pass.

**Comment by bjorn3** [comment]
> (But to my knowledge, at least on Apple's aarch64 targets, that is actually unsound.)

Indeed. Variadic arguments are always passed on the stack even if the same argument would end up in a register for a non-variadic function.

**Comment by boozook** [comment]
👋🏻 Is there any progress on c-variadics? I need this and would love to help.

**Comment by RalfJung** [comment]
The underlying blocker has been resolved long ago, but a bunch of work on the core interpreter is still required to support variadic calls:
- The variadic arguments need to be stored *somewhere*.
- The `va_*` intrinsics need to be wired up with wherever those arguments are stored.
- We need to thoroughly think through all the ways one can cause UB with variadics, and detect them all.

Most of this work will have to happen in the main rust repository where the core interpreter lives. I can provide some general guidance, but won't have the time to provide a lot of in-depth mentoring, so whoever takes this on should be comfortable with working fairly independently on a large and complicated codebase.
